plugin/pkg/admission/podtolerationrestriction/admission.go

```diff
@@ -0,0 +1,221 @@
+*/
+
+package podtolerationrestriction
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	informers "k8s.io/kubernetes/pkg/client/informers/informers_generated/internalversion"
+	corelisters "k8s.io/kubernetes/pkg/client/listers/core/internalversion"
+	kubeapiserveradmission "k8s.io/kubernetes/pkg/kubeapiserver/admission"
+	"k8s.io/kubernetes/pkg/util/tolerations"
+	pluginapi "k8s.io/kubernetes/plugin/pkg/admission/podtolerationrestriction/apis/podtolerationrestriction"
+)
+
+func init() {
+	admission.RegisterPlugin("PodTolerationRestriction", func(config io.Reader) (admission.Interface, error) {
+		pluginConfig, err := loadConfiguration(config)
+		if err != nil {
+			return nil, err
+		}
+		return NewPodTolerationsPlugin(pluginConfig), nil
+	})
+}
+
+// The annotation keys for default and whitelist of tolerations
+const (
+	NSDefaultTolerations string = "scheduler.alpha.kubernetes.io/defaultTolerations"
+	NSWLTolerations      string = "scheduler.alpha.kubernetes.io/tolerationsWhitelist"
+)
+
+var _ = kubeapiserveradmission.WantsInformerFactory(&podTolerationsPlugin{})
+
+type podTolerationsPlugin struct {
+	*admission.Handler
+	client          clientset.Interface
+	namespaceLister corelisters.NamespaceLister
+	pluginConfig    *pluginapi.Configuration
+}
+
+// This plugin first verifies any conflict between a pod's tolerations and
+// its namespace's tolerations, and rejects the pod if there's a conflict.
+// If there's no conflict, the pod's tolerations are merged with its namespace's
+// toleration. Resulting pod's tolerations are verified against its namespace's
+// whitelist of tolerations. If the verification is successful, the pod is admitted
+// otherwise rejected. If a namespace does not have associated default or whitelist
+// of tolerations, then cluster level default or whitelist of tolerations are used
+// instead if specified. Tolerations to a namespace are assigned via
+// scheduler.alpha.kubernetes.io/defaultTolerations and scheduler.alpha.kubernetes.io/tolerationsWhitelist
+// annotations keys.
+func (p *podTolerationsPlugin) Admit(a admission.Attributes) error {
+	resource := a.GetResource().GroupResource()
+	if resource != api.Resource("pods") {
+		return nil
+	}
+	if a.GetSubresource() != "" {
+		// only run the checks below on pods proper and not subresources
+		return nil
+	}
+
+	obj := a.GetObject()
+	pod, ok := obj.(*api.Pod)
+	if !ok {
+		glog.Errorf("expected pod but got %s", a.GetKind().Kind)
+		return nil
+	}
+
+	if !p.WaitForReady() {
+		return admission.NewForbidden(a, fmt.Errorf("not yet ready to handle request"))
+	}
+
+	nsName := a.GetNamespace()
+	namespace, err := p.namespaceLister.Get(nsName)
+	if errors.IsNotFound(err) {
```

Comment:
Is it better like this?
```
	if err != nil {
		if errors.IsNotFound(err) {
			// in case of latency in our caches, make a call direct to storage to verify that it truly exists or not
			namespace, err = p.client.Core().Namespaces().Get(nsName, metav1.GetOptions{})
			if err != nil {
				if errors.IsNotFound(err) {
					return err
				}
				return errors.NewInternalError(err)
			}
		}
		return errors.NewInternalError(err)
	}
```